rbd: add backend support for VolumeGroup operations

nixpanic opened the pull request:
# Describe what this PR does #

Implementation of the VolumeGroup backend. It contains a clear separation of the VolumeGroup type and functions and CSI-Addons responsibilities.

## Is there anything that requires special attention ##

Tested with a [small application that calls the CSI-Addons procedures directly](https://gist.github.com/nixpanic/3aab1eb6ad1937a52fb289e91ace3272). The application was modified several times to test different scenarios. At the moment I am confident that the basic functionality works as expected.

```
I0722 14:28:27.040708       1 utils.go:235] ID: 22 GRPC call: /volumegroup.Controller/CreateVolumeGroup
I0722 14:28:27.040980       1 utils.go:236] ID: 22 GRPC request: {"name":"my-group","parameters":{"clusterID":"openshift-storage","pool":"ocs-storagecluster-cephblockpool"},"secrets":"***stripped***","volume_ids":["0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88","0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43"]}
I0722 14:28:27.065729       1 omap.go:89] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.eafd9e22-70de-459b-97fc-2d46e03d5f88"): map[csi.imageid:67c16fdb4f5d csi.imagename:csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88 csi.volname:pvc-37249c71-61f2-4c2f-8cdd-08cb2f99d668 csi.volume.owner:default]
I0722 14:28:27.112579       1 omap.go:89] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.2f1acca8-95ca-4594-be0c-e5e70b6cbc43"): map[csi.imageid:67c1ec38b881 csi.imagename:csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43 csi.volname:pvc-5aae7b38-7ec3-4c45-b3c9-1a21c1d37b78 csi.volume.owner:default]
I0722 14:28:27.143738       1 volumegroup.go:106] ID: 22 all 2 Volumes for VolumeGroup "my-group" have been found
I0722 14:28:27.156146       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.groups."): map[csi.volume.group.my-group:46ae4226-e71b-4955-8440-4801f178979a])
I0722 14:28:27.160879       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group])
I0722 14:28:27.161611       1 omap.go:221] ID: 22 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group]
I0722 14:28:27.161663       1 volume_group.go:149] ID: 22 GetVolumeGroup(0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a) returns {id:0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a name:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a clusterID:openshift-storage credentials:0xc0008111a0 conn:<nil> ioctx:<nil> monitors:172.30.253.8:3300,172.30.149.93:3300,172.30.68.64:3300 pool:ocs-storagecluster-cephblockpool namespace: journal:0xc000b2e840 volumes:[] volumesToFree:[]}
I0722 14:28:27.161697       1 volume_group.go:249] ID: 22 connection established for volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
I0722 14:28:27.161710       1 volume_group.go:277] ID: 22 iocontext created for volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" in pool "ocs-storagecluster-cephblockpool"
I0722 14:28:27.170848       1 volume_group.go:326] ID: 22 volume group "csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a" has been created
I0722 14:28:27.170869       1 volumegroup.go:118] ID: 22 VolumeGroup "my-group" had been created: csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a
I0722 14:28:27.194642       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88:])
I0722 14:28:27.216181       1 omap.go:159] ID: 22 set omap keys (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43:])
I0722 14:28:27.216203       1 volumegroup.go:133] ID: 22 all 2 Volumes have been added to for VolumeGroup "my-group"
I0722 14:28:27.216519       1 utils.go:242] ID: 22 GRPC response: {"volume_group":{"volume_group_id":"0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a","volumes":[{"capacity_bytes":17179869184,"volume_context":{"imageName":"csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88","journalPool":"ocs-storagecluster-cephblockpool","pool":"ocs-storagecluster-cephblockpool"},"volume_id":"0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88"},{"capacity_bytes":8589934592,"volume_context":{"imageName":"csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43","journalPool":"ocs-storagecluster-cephblockpool","pool":"ocs-storagecluster-cephblockpool"},"volume_id":"0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43"}]}}
I0722 14:28:27.217582       1 utils.go:235] ID: 23 GRPC call: /volumegroup.Controller/DeleteVolumeGroup
I0722 14:28:27.217667       1 utils.go:236] ID: 23 GRPC request: {"secrets":"***stripped***","volume_group_id":"0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"}
I0722 14:28:27.218503       1 omap.go:221] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.group.46ae4226-e71b-4955-8440-4801f178979a"): map[0001-0011-openshift-storage-0000000000000001-2f1acca8-95ca-4594-be0c-e5e70b6cbc43: 0001-0011-openshift-storage-0000000000000001-eafd9e22-70de-459b-97fc-2d46e03d5f88: csi.groupname:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a csi.volname:my-group]
I0722 14:28:27.219984       1 omap.go:89] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.2f1acca8-95ca-4594-be0c-e5e70b6cbc43"): map[csi.imageid:67c1ec38b881 csi.imagename:csi-vol-2f1acca8-95ca-4594-be0c-e5e70b6cbc43 csi.volname:pvc-5aae7b38-7ec3-4c45-b3c9-1a21c1d37b78 csi.volume.owner:default]
I0722 14:28:27.258472       1 omap.go:89] ID: 23 got omap values: (pool="ocs-storagecluster-cephblockpool", namespace="", name="csi.volume.eafd9e22-70de-459b-97fc-2d46e03d5f88"): map[csi.imageid:67c16fdb4f5d csi.imagename:csi-vol-eafd9e22-70de-459b-97fc-2d46e03d5f88 csi.volname:pvc-37249c71-61f2-4c2f-8cdd-08cb2f99d668 csi.volume.owner:default]
I0722 14:28:27.289357       1 volume_group.go:149] ID: 23 GetVolumeGroup(0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a) returns {id:0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a name:csi-vol-group-46ae4226-e71b-4955-8440-4801f178979a clusterID:openshift-storage credentials:0xc000c832c0 conn:<nil> ioctx:<nil> monitors:172.30.253.8:3300,172.30.149.93:3300,172.30.68.64:3300 pool:ocs-storagecluster-cephblockpool namespace: journal:0xc00016f640 volumes:[0xc000658b48 0xc000658d88] volumesToFree:[0xc000658b48 0xc000658d88]}
I0722 14:28:27.289378       1 volumegroup.go:185] ID: 23 VolumeGroup "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" has been found
I0722 14:28:27.289387       1 volumegroup.go:197] ID: 23 VolumeGroup "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a" contains 2 volumes
I0722 14:28:27.289420       1 volume_group.go:307] ID: 23 destroyed volume group instance with id "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
E0722 14:28:27.289438       1 utils.go:240] ID: 23 GRPC error: rpc error: code = FailedPrecondition desc = rejecting to delete non-empty volume group "0001-0011-openshift-storage-0000000000000001-46ae4226-e71b-4955-8440-4801f178979a"
```

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

nixpanic committed:
cleanup: make VolumeGroupJournalConnection a private type

VolumeGroupJournalConnection is not used outside the internal/journal
package. There is no need to expose the type outside of the package, it
causes only confusion about the usage of the journalling API.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: pass CSI-instanceID to CSI-Addons VolumeGroupServer

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: update Volume interface implementation for VolumeGroup APIs

Add support for adding and removing the RBD-image from a group.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 commented on internal/rbd/controllerserver.go:
```diff
@@ -266,15 +266,19 @@ func (rbdVol *rbdVolume) ToCSI(ctx context.Context) *csi.Volume {
-	volume := rbdVol.ToCSI(ctx)
+	volume, err := rbdVol.ToCSI(ctx)
+	if err != nil {
+		log.ErrorLog(ctx, "BUG, can not happen: failed to convert volume %q to CSI type: %w", rbdVol, err)
```
panic or return error here as well? continue will result in other problems

nixpanic replied:
Do you have a preference to panic or return an error? I guess returning an error is the cleanest.

Madhu-1 replied:
return error looks cleaner

nixpanic replied:
done

Madhu-1 commented on internal/rbd/group.go:
```diff
@@ -0,0 +1,58 @@
+	"context"
+	"fmt"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+
+	"github.com/ceph/ceph-csi/internal/rbd/types"
```
rearrange the order here? cephcsi import should come first

nixpanic replied:
this guideline is not followed everywere. I think and Golang suggests to have imports of local packages at the bottom, that seems to be a coding convention many other Go projects use.

Madhu-1 replied:
https://github.com/ceph/ceph-csi/blob/devel/docs/coding.md#imports this is one we have, lets stick to it and also update other places if its not followed as followup cleanup?

nixpanic replied:
See #4721. I don't think we should strictly enforce it for existing imports, but it would be nice to use this order in new files.

Madhu-1 commented on internal/rbd/group.go:
```diff
@@ -0,0 +1,58 @@
+	return librbd.GroupImageAdd(ioctx, name, rv.ioctx, rv.RbdImageName)
+}
+
+// RemoveFromGroup removes the image to the group. This is called from the
```
`image to the group` to `image from the group`

Madhu-1 commented on internal/rbd/group.go:
```diff
@@ -0,0 +1,58 @@
+// RemoveFromGroup removes the image to the group. This is called from the
+// rbd_group package.
+func (rv *rbdVolume) RemoveFromGroup(ctx context.Context, vg types.VolumeGroup) error {
+	ioctx, err := vg.GetIOContext(ctx)
```
ioctx is not used

nixpanic replied:
yes it is, on `return librbd.GroupImageAdd(ioctx, ...`.

Madhu-1 commented on internal/rbd/group.go:
```diff
@@ -0,0 +1,58 @@
+// AddToGroup adds the image to the group. This is called from the rbd_group
+// package.
+func (rv *rbdVolume) AddToGroup(ctx context.Context, vg types.VolumeGroup) error {
+	ioctx, err := vg.GetIOContext(ctx)
```
ioctx is not used

nixpanic replied:
yes it is, on `return librbd.GroupImageAdd(ioctx, ...`.

Madhu-1 replied:
but you are passing rv.ioctx not the one you get from here

Madhu-1 replied:
ha my mistake, it looks like we have two ioctx

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -47,43 +53,166 @@ func (mgr *rbdManager) Destroy(ctx context.Context) {
+	if err != nil {
+		return nil, err
+	}
+
```
i think in case of any error we need to destroy the connection to journal here

nixpanic replied:
The Manager takes care of that during `Destroy()`. This makes the code simpler as there are no `defer` statements needed that check for errors.

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -47,43 +53,166 @@ func (mgr *rbdManager) Destroy(ctx context.Context) {
+	if err != nil {
+		return nil, err
+	}
+
```
same here as well Destory is required?

nixpanic replied:
no, see above.

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -47,43 +53,166 @@ func (mgr *rbdManager) Destroy(ctx context.Context) {
+	pool, ok := mgr.parameters["pool"]
+	if !ok {
+		return nil, errors.New("required 'pool' parameter missing in volume attributes")
+	}
```
validate `pool` is not empty string?

Madhu-1 commented on internal/rbd_group/volume_group.go:
```diff
@@ -0,0 +1,340 @@
+	"errors"
+	"fmt"
+
+	"github.com/ceph/go-ceph/rados"
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/csi-addons/spec/lib/go/volumegroup"
+
+	"github.com/ceph/ceph-csi/internal/journal"
+	"github.com/ceph/ceph-csi/internal/rbd/types"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
```
reorder is required here

nixpanic replied:
I don't like that :wink: , many Go projects prefer this ordering. It would be nice if Ceph-CSI can use a similar standard formatting.

nixpanic committed:
rbd: implement the VolumeGroup interface

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: add journalledObject as base for VolumeGroup interface

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: remove the VolumeGroup from the journal on DeleteVolumeGroup

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 reviewed (commented):
LGTM, let me know once its tested, i will do final review and approve it.

nixpanic committed:
rbd: make VolumeGroup Create/Delete/AddVolume/RemoveVolume idempotent

Add extra error checking to make sure trying to create an existing
volume group does not result in a failure. The same counts for deleting
a non-existing volume group, and adding/removing volumes to/from the
volume group.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: use the Manager to handle CSI-Addons VolumeGroup requests

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: check if an image is part of a group before adding it

A RBD image can only be part of a single group. While an image is added
to a group, check if the image is already part of a group, and return an
error in case it is.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

yati1998 commented:
LGTM

Rakshith-R commented on internal/rbd/manager.go:
```diff
@@ -47,43 +61,211 @@ func (mgr *rbdManager) Destroy(ctx context.Context) {
+	// volumeNamePrefix is an optional parameter, can be an empty string
+	prefix := mgr.parameters["volumeNamePrefix"]
+
+	uuid, vgName, err := vgJournal.ReserveName(ctx, journalPool, name, prefix)
```
https://github.com/ceph/ceph-csi/blob/devel/internal/journal/volumegroupjournal.go#L170 

checkreservation needs to be called before trying to reserve the name ? 

similar to https://github.com/ceph/ceph-csi/blob/devel/internal/cephfs/store/volumegroup.go#L199

nixpanic replied:
this is done now, it seems to be working well with the added `CheckReservation` call.

Rakshith-R commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -115,7 +115,7 @@ func (vs *VolumeGroupServer) CreateVolumeGroup(
 	}
 
-	log.DebugLog(ctx, fmt.Sprintf("VolumeGroup %q had been created", req.GetName()))
+	log.DebugLog(ctx, fmt.Sprintf("VolumeGroup %q had been created: %+v", req.GetName(), vg))
```
```
	log.DebugLog(ctx, fmt.Sprintf("VolumeGroup %q has been created: %+v", req.GetName(), vg))
Comment
```

nixpanic replied:
@Rakshith-R, sorry, I don't understand what you mean by this?

The `fmt.Sprintf()` call can be dropped, is that it?

Rakshith-R reviewed (commented):
Tiny nits.

There's 3 commits named `rbd: implement the VolumeGroup interface` that I think can be squashed together.

Rakshith-R reviewed (approved):
Thanks LGTM

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
@mergifyio requeue

The **multi-arch-build** CI job failed while installing some packages. It has been restarted and has passed the section that previously failed.
